### PR TITLE
Add an example of JWT authorization

### PR DESCRIPTION
**Pull request by musaprg**
Signed-off-by: Kotaro Inoue <k.musaino@gmail.com>

This PR adds a simple example of JWT authorization.
The JWT token used in this example is expected to be signed in HMAC256 (HMAC-SHA256).

### CODE REVIEW

File: examples/http_jwt_auth/main.go [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+		return types.ActionPause
+	}
+
+	proxywasm.LogInfof("request authorized!")
+
+	return types.ActionContinue
+}
+
+// verifyToken checks if the JWT token is valid.
+func verifyToken(token string) (bool, error) {
```
Comment (mathetake) [inline examples/http_jwt_auth/main.go:105]: bool seems redundant. So just return an error  if the validation failed
```suggestion
func verifyToken(token string) error {
```

File: examples/http_jwt_auth/main.go [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+	if err != nil {
+		if err := proxywasm.SendHttpResponse(400, nil, []byte("authorization header must be provided"), -1); err != nil {
+			panic(err)
+		}
+		return types.ActionPause
+	}
+
+	proxywasm.LogInfof("authorization token: %s", authorization)
+
+	// Validate header format
```
Comment (mathetake) [inline examples/http_jwt_auth/main.go:78]: Always make sure comments start with capital letters and end with period.
```suggestion
	// Validate header format.
```

File: examples/http_jwt_auth/main.go [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+	slice := strings.Fields(authorization)
+	if len(slice) != 2 || slice[0] != "Bearer" {
+		if err := proxywasm.SendHttpResponse(400, nil, []byte("invalid authorization header"), -1); err != nil {
+			panic(err)
+		}
+		return types.ActionPause
+	}
+
+	// Verify token
```
Comment (mathetake) [inline examples/http_jwt_auth/main.go:87]:
```suggestion
	// Verify token.
```

File: examples/http_jwt_auth/main.go [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+
+// verifyToken checks if the JWT token is valid.
+func verifyToken(token string) (bool, error) {
+	sepIdx := strings.LastIndex(token, ".")
+	if sepIdx == -1 {
+		return false, nil
+	}
+	unsignedToken := token[:sepIdx]
+	signature, err := base64.RawURLEncoding.DecodeString(token[sepIdx+1:])
+	if err != nil { // invalid base64 data
```
Comment (mathetake) [inline examples/http_jwt_auth/main.go:112]: unnecessary comment.
```suggestion
	if err != nil {
```

File: examples/http_jwt_auth/main_test.go [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+	"testing"
+
+	"github.com/stretchr/testify/require"
+
+	"github.com/tetratelabs/proxy-wasm-go-sdk/proxywasm/proxytest"
+	"github.com/tetratelabs/proxy-wasm-go-sdk/proxywasm/types"
+)
+
+const (
+	// sampleJwtToken is a sample JWT token which is already signed with secret key "secret".
```
Comment (mathetake) [inline examples/http_jwt_auth/main_test.go:22]:
```suggestion
	// A sample JWT token which is already signed with secret key "secret".
```

File: examples/http_jwt_auth/main_test.go [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+				// Must be paused.
+				require.Equal(t, types.ActionPause, action)
+
+				// Check the local response.
+				localResponse := host.GetSentLocalResponse(id)
+				require.NotNil(t, localResponse)
+				require.Equal(t, uint32(400), localResponse.StatusCode)
+				require.Equal(t, "invalid authorization header", string(localResponse.Data))
+
+				// Call OnHttpStreamDone
```
Comment (mathetake) [inline examples/http_jwt_auth/main_test.go:97]:
```suggestion
				// Call OnHttpStreamDone.
```

File: examples/http_jwt_auth/main_test.go [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+				// Must be paused.
+				require.Equal(t, types.ActionPause, action)
+
+				// Check the local response.
+				localResponse := host.GetSentLocalResponse(id)
+				require.NotNil(t, localResponse)
+				require.Equal(t, uint32(401), localResponse.StatusCode)
+				require.Equal(t, "invalid token", string(localResponse.Data))
+
+				// Call OnHttpStreamDone
```
Comment (mathetake) [inline examples/http_jwt_auth/main_test.go:134]:
```suggestion
				// Call OnHttpStreamDone.
```

File: examples/http_jwt_auth/main_test.go [hunk 8]
```diff
@@ -0,0 +1,139 @@
...
+		// Must be paused.
+		require.Equal(t, types.ActionPause, action)
+
+		// Check the local response.
+		localResponse := host.GetSentLocalResponse(id)
+		require.NotNil(t, localResponse)
+		require.Equal(t, uint32(400), localResponse.StatusCode)
+		require.Equal(t, "authorization header must be provided", string(localResponse.Data))
+
+		// Call OnHttpStreamDone
```
Comment (mathetake) [inline examples/http_jwt_auth/main_test.go:63]:
```suggestion
		// Call OnHttpStreamDone.
```

File: examples/http_jwt_auth/main_test.go [hunk 9]
```diff
@@ -0,0 +1,139 @@
...
+
+		// Call OnRequestHeaders.
+		action := host.CallOnRequestHeaders(id, [][2]string{
+			{"Authorization", fmt.Sprintf("Bearer %s", sampleJwtToken)},
+		}, false)
+
+		// Must be continued.
+		require.Equal(t, types.ActionContinue, action)
+
+		// Call OnHttpStreamDone
```
Comment (mathetake) [inline examples/http_jwt_auth/main_test.go:43]:
```suggestion
		// Call OnHttpStreamDone.
```

File: examples/http_jwt_auth/main.go [hunk 10]
```diff
@@ -0,0 +1,119 @@
...
+func (ctx *httpContext) OnHttpRequestHeaders(numHeaders int, endOfStream bool) types.Action {
+	authorization, err := proxywasm.GetHttpRequestHeader("Authorization")
+	if err != nil {
+		if err := proxywasm.SendHttpResponse(400, nil, []byte("authorization header must be provided"), -1); err != nil {
+			panic(err)
+		}
+		return types.ActionPause
+	}
+
+	proxywasm.LogInfof("authorization token: %s", authorization)
```
Comment (mathetake) [inline examples/http_jwt_auth/main.go:76]: I imagine this should not be logged?

**Review by mathetake — CHANGES_REQUESTED** [review]

File: examples/http_jwt_auth/main_test.go [hunk 11]
```diff
@@ -0,0 +1,139 @@
...
+	t.Run("401 response for invalid token", func(t *testing.T) {
+		testCases := []struct {
+			name  string
+			token string
+		}{
+			{"invalid token with no dot", "invalidtoken"},
+			{"invalid token with one dot", "invalid.token"},
+			{"invalid token with two dot", "invalid.token.hoge"},
+			{"invalid token with non-base64 encoded signature", "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.e30.hogefugafoo"},
+			{"invalid token with base64 encoded signature", "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.e30.aG9nZWZ1Z2Fmb28K"},
```
Comment (mathetake) [inline examples/http_jwt_auth/main_test.go:112]: they seem redundant prefix to me
```suggestion
			{"no dot", "invalidtoken"},
			{"one dot", "invalid.token"},
			{"two dot", "invalid.token.hoge"},
			{"non-base64 encoded signature", "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.e30.hogefugafoo"},
			{"base64 encoded signature", "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.e30.aG9nZWZ1Z2Fmb28K"},
```

File: examples/http_jwt_auth/envoy.yaml [hunk 12]
```diff
@@ -0,0 +1,89 @@
...
+                      value:
+                        config:
+                          vm_config:
+                            runtime: "envoy.wasm.runtime.v8"
+                            code:
+                              local:
+                                filename: "./examples/http_jwt_auth/main.go.wasm"
+                  - name: envoy.filters.http.router
+
+    - name: staticreply
```
Comment (mathetake) [inline examples/http_jwt_auth/envoy.yaml:40]: do we need to have staticreply in the first place? Having the direct response in the router seems enough to me for this example: https://github.com/tetratelabs/proxy-wasm-go-sdk/blob/main/examples/vm_plugin_configuration/envoy.yaml#L24-L27

**Review by mathetake — COMMENTED** [review]
another  ones

File: examples/http_jwt_auth/main.go (hunk 2 above)
Comment (musaprg) [reply examples/http_jwt_auth/main.go:78]: OK

File: examples/http_jwt_auth/envoy.yaml (hunk 12 above)
Comment (musaprg) [reply examples/http_jwt_auth/envoy.yaml:40]: That's a better idea. OK, I will change to use it.

**Comment by mathetake** [comment]
talked offline, and I'm closing this issue as @musaprg won't be able to work on this anymore.

**Pull request closed without merging**